tests/test_0074-argsort-and-sort.py

```diff
@@ -368,15 +466,25 @@ def test_RecordArray():
         [],
     ]
     assert ak.to_list(array.y.layout.argsort(0, True, False)) == [
+        # FIXME?
         [],
-        [0],
-        [1, 0],
-        [2, 1, 0],
-        [3, 2, 1, 0],
-        [4, 3, 2],
-        [5, 4],
-        [6],
-        [],
+        [1],
+        [2, 2],
+        [3, 3, 3],
+        [4, 4, 4, 4],
+        [5, 5, 5],
+        [6, 6],
+        [7],
+        []
+        # [],
```

Comment:
@jpivarski - this result at `axis=0` looks a bit odd to me. Though I have a feeling it it correct. It's more obvious if I modify the array to:
```python
>>> array = ak.Array([[], [-1], [2, 0.2], [0.3, 3.3, 3], [4, 4.4, 0.4, -4], [-5, 5, 5.5], [0.6, 6], [-7], []])
>>> ak.argsort(array, axis=0)
<Array [[], [7], [5, 2], ... [2, 6], [4], []] type='9 * var * int64'>
>>> ak.sort(array, axis=0)
<Array [[], [-7], [-5, 0.2, ... 2, 6], [4], []] type='9 * var * float64'>
```
and then try to access values by an argsorted index individually. They do correlate with the same values from a sorted array.

BTW, I can use the `argsort` at `axis=-1` to sort the initial array in the same axis, but not if I use `axis=0`:
```python
>>> array[ak.argsort(array)]
<Array [[], [-1], [0.2, 2, ... 6], [-7], []] type='9 * var * float64'>
>>> array[ak.argsort(array, axis=0)]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/yana/Projects/PR945/awkward-1.0/awkward/highlevel.py", line 993, in __getitem__
    tmp = ak._util.wrap(self.layout[where], self._behavior)
ValueError: in ListArray64 attempting to get 7, index out of range

(https://github.com/scikit-hep/awkward-1.0/blob/1.4.0rc2/src/cpu-kernels/awkward_ListArray_getitem_jagged_apply.cpp#L46)
```
Is it a feature or a bug? Thanks!

Reply:
This is the important thing:

```python
>>> ak.all(array[ak.argsort(array, axis=-1)] == ak.sort(array, axis=-1))
True
```

As for `axis=0`, the meaning of sorting with `axis=0` cuts across lists:

```python
>>> ak.sort(array, axis=0).tolist()
[[], [-7.0], [-5.0, 0.2], [-1.0, 3.3, 0.4], [0.3, 4.4, 3.0, -4.0], [0.6, 5.0, 5.5], [2.0, 6.0], [4.0], []]
```

That is, `-7.0` is the smallest number that is first in any list, `-5.0` is the second smallest number that is first in any list, and `0.2` is the smallest number that is second in any list. I don't know that it makes sense to maintain the shapes of all these lists when you're doing that, and I don't know what analysts would _want_ it to mean in the context of a data analysis.

At least, it's no mystery that `array[ak.argsort(array, axis=0)]` fails: the index positions returned by `ak.argsort` cut across all of the lists—they can be values that are much larger than the lengths of any of the lists.

```python
>>> ak.argsort(array, axis=0).tolist()
[[], [7], [5, 2], [1, 3, 4], [3, 4, 3, 4], [6, 5, 5], [2, 6], [4], []]
```

So what should we do about it? Maybe—without removing any of the the code that works—we should put in an exception that refuses to return results when `axis != -1` (or the equivalent positive `axis`, so this has to be in the C++ code, right after `posaxis` is determined from `wrap_axis_if_negative`). Then we're not promising that this is the behavior, such that someone can start relying on it before we know what the right policy should be. Also, if someone really does want `axis != -1 and axis is not None`, they'll complain that it's not working and we can ask them about their use-case and understand what the appropriate output would be.

What do you think?

Reply:
btw, NumPy handles it in a weird way:

```python
>>> a = np.array([[1,4],[3,1]])
>>> np.sort(a, axis=0)
array([[1, 1],
       [3, 4]])
>>> np.argsort(a, axis=0)
array([[0, 1],
       [1, 0]])
```
but
```python
>>> a[np.argsort(a, axis=0)]
array([[[1, 4],
        [3, 1]],

       [[3, 1],
        [1, 4]]])
```
The same in axis=-1:
```python
>>> np.sort(a, axis=-1)
array([[1, 4],
       [1, 3]])
>>> np.argsort(a, axis=-1)
array([[0, 1],
       [1, 0]])
>>> a[np.argsort(a, axis=-1)]
array([[[1, 4],
        [3, 1]],

       [[3, 1],
        [1, 4]]])
```